### DISCUSSION TITLE
Confusion with lookup tables

### DISCUSSION THREAD

**Discussion by front-endian** [Myndex/SAPC-APCA#64] [Q&A]
Howdy,

We've continued trying to incorporate APCA into our internal design guidelines, starting with a project to create a new shared design system across many products (web, desktop, and mobile apps).

One bit of confusion that has come up is that the LUTs current only separates body text from non-body content in a handful of spots:

<img width="318" alt="Screen Shot 2022-03-18 at 5 35 52 PM" src="https://user-images.githubusercontent.com/11234702/159086509-07f42b85-8c3e-4f9d-909b-f6c3a8dd647c.png">

Some recommendations (such as [https://github.com/Myndex/SAPC-APCA/discussions/31](https://github.com/Myndex/SAPC-APCA/discussions/31)) split contrast requirements completely between different content types completely. That flexibility is great to have, but without a full LUT, it is less useful. This is especially true as we make some very data dense applications that need to use non-body text sizes around 14–16px.

There are a lot of questions I have around this, and feedback for discussions you have open like [https://github.com/Myndex/SAPC-APCA/discussions/45](https://github.com/Myndex/SAPC-APCA/discussions/45), but I'll start with these questions:

- For normal weight text at 14–16px text the current APCA Contrast LUT only gives a body text value of Lc 85–95. Can non-body text go to lower levels?
- Do we subtract Lc 15 to get the non-body contrast requirement for these smaller sizes?
- Why is it the LUT only splits out body from non-body text in some cells and not others?

Once we get answers to these I can dig in further with some other things we've found and share some feedback from having a lot of designers look at the current APCA guidelines and LUTs.

**Comment by Myndex** [comment]
Hi Colin @justcolin 

> _One bit of confusion that has come up is that the LUTs current only separates body text from non-body content in a handful of spots:_

This is a good question, and we are considering a 3D LUT to address this, but in the meantime, let me give you some additional guidance. There was an earlier multi level type LUT displayed at the SAPC development site, but for APCA W3 we have been working on simplification.

> _Some recommendations (such as #31) split contrast requirements completely between different content types completely. That flexibility is great to have, but without a full LUT, it is less useful. This is especially true as we make some very data dense applications that need to use non-body text sizes around 14–16px._

Can you show me an example? If it's proprietary, you can email it to me in confidence — there is discussion regarding divisions of use cases, best fluent, fluent, spot, etc etc...

-----
## _Answers_
> * _For normal weight text at 14–16px text the current APCA Contrast LUT only gives a body text value of Lc 85–95. Can non-body text go to lower levels?_
> * _Do we subtract Lc 15 to get the non-body contrast requirement for these smaller sizes?_

If the use case is "spot" non-content then yes, the amount of derating to be determined, but probably Lc 15 to 20 ish, depending on size and weight. ("spot" covers non-content text such as copyright notices, disabled links or controls, placeholder text, etc).

If the use case is _fluent_ content that is not body text, then mostly no, but again it is weight dependent. At issue is the major stroke width, as referenced in a capital I.

We are discussing an additional use-case for "soft content" that is a level stronger than spot, but not at the full needs of fluent text, for things like call-outs, legend elements, etc.


> * _Why is it the LUT only splits out body from non-body text in some cells and not others?_

Okay, so this gets directly into how fonts are rasterized on a typical (not retina) display. Even with subpixel antialiasing, small thin fonts tend to render at a softened and sometimes lower contrast than the CSS color would imply. This is sometimes caused by improper use of the "-webkit-font-smoothing:" property, or by some rasterizers, or other reasons.

Small font sizes and weights where this is likely to be an issue, are set to a minimum Lc value that ends up being close to the value needed for body text, so it is set as "green" for body text.

Larger heavier fonts that are not prone to this problem can be at a lower Lc for single line type applications, and still be "fluently readable" .... but if put into a block of body text (which reduces perceived readability contrast) then you need to add Lc 15 for body text. And this is the reason for the split.

The small fonts that have high contrast requirements for rasterizing reasons are already at the high level needed for body blocks, but they would be good to have Lc 15 added AS WELL... however, that would also become restrictive for design.

-----
### _Some Foundational Basics_
For dark text on light backgrounds, and the text is 24px or smaller, the text should be ` #000000 ` 

Black is _ideal_ for dark text, and really the only time dark text on a light background should be anything other than black is when:

1) You don't want anyone to be able to read it. (joke)
2) The text is _larger_ than 24px bold or 36px normal.
3) It is blue, to indicate an inline link.

With black text ` #000 ` you can use a nice relaxed background like ` #e8e4d8 ` and still achieve Lc 90.

On the subject of a blue link: with the ` #e8e4d8 ` adding blue to the black like ` #0000aa ` would lower the Lc value and indicate a larger font, though we're working on a blue exception, to permit ` #00a ` to calculate as ` #000 ` for short periods (i.e. link use) as blue is not a significant part of readability luminance contrast.

I am not certain where the "grey text" thing came from, but it's not good. I suspect it is a misunderstanding on using "reduced contrast" but reducing contrast for an element with dark text on a light background is best done by darkening the background.


-----
> _Once we get answers to these I can dig in further with some other things we've found and share some feedback from having a lot of designers look at the current APCA guidelines and LUTs._

Excellent, I am looking forward to hearing comments. We have a lot of potential options and considerations that are not yet released, waiting to hear from beta testers using the current "most simple" versions.

**Reply by front-endian** [reply]
Thanks for getting back! Took a bit of time to get some examples of the sorts of websites I regularly deal with to better illustrate what we deal with.

> This is a good question, and we are considering a 3D LUT to address this, but in the meantime, let me give you some additional guidance. There was an earlier multi level type LUT displayed at the SAPC development site, but for APCA W3 we have been working on simplification.

Can you describe in a bit more detail what the 3D LUT would be? One area I am especially interested in is a LUT for non-text elements. A LUT to determine what is the acceptable contrast for a given line/element thickness would be extremely helpful for creating accessible data visualizations, non-text interactive elements etc.

> Can you show me an example? If it's proprietary, you can email it to me in confidence — there is discussion regarding divisions of use cases, best fluent, fluent, spot, etc etc...

I do a lot of work with admin interfaces used by networking and security professionals. In those applications, large data tables with links/buttons are extremely common. Due to the nature of how tables scale and the sorts of data we have to display there are limits on what we can do in terms of text sizes. I've spoken with some of our senior designers, played with our UI's a fair amount, and the limits are around 14–16px. Any larger than that and event some pretty tame data tables start breaking at reasonable screen sizes. (Even getting up to 14px won't be easy, as many have spend years doing designs around 10–12px, and changing those habits takes time.)

The following examples are just illustrative. Yes, they have many accessibility problems, yes links need to be underlined, colors changed, etc. These examples come from a variety of applications, some of which I don't work on, some which are currently undergoing redesigns, etc. Sharing them not as examples of good accessibility, but the sorts of data tables commonly used for IT admin interfaces:

<img width="1017" alt="Screen Shot 2022-03-21 at 4 10 25 PM" src="https://user-images.githubusercontent.com/11234702/159565675-6a430d49-3d63-4f06-8cfa-f55724f4dcf2.png">
<img width="1349" alt="Screen Shot 2022-03-21 at 4 08 54 PM" src="https://user-images.githubusercontent.com/11234702/159565710-1549b691-00b0-4636-90c8-7304ec33ceec.png">
<img width="1144" alt="Screen Shot 2022-03-21 at 4 06 36 PM" src="https://user-images.githubusercontent.com/11234702/159565758-cc5cfb7a-6c5a-4cef-b358-cc204f5d18dc.png">
<img width="719" alt="Screen Shot 2022-03-22 at 3 25 16 PM" src="https://user-images.githubusercontent.com/11234702/159565785-4a070f7a-24ae-4612-a117-6074eb583d01.png">
<img width="573" alt="Screen Shot 2022-03-22 at 4 14 06 PM" src="https://user-images.githubusercontent.com/11234702/159567918-112790fc-42b8-40f0-8920-fcefe388b158.png">

Some tables will get even more information dense than this, but they should suffice as average examples.

Getting plain text in data tables to be upwards of Lc 95 for these tables is pretty easy. However, there is no way to make embedded interactive elements like normal weight links Lc 95 without having a link color that is almost indistinguishable from the plain text color. With the current LUT, we could consider making those links bold, dropping the requirement down to Lc 75. However, I question whether large paragraphs of body text really needs to have the same contrast requirements as a button or link with only one or two words in it.

It would be very useful to have "soft content" requirements for these situations with short bits of text (labels, links/buttons in tables, menu bars, etc) where there are multiple design requirements and we can't just have arbitrarily high contrast without losing out on some other element of usability. However, you know more about the science and thus my want to have a discussion with you. 😁

**Note:** specific color recommendations are not helpful given the sorts of very large scale projects and color systems I advise on. My interest is not in what specific colors will/won't work in specific situations, but understanding the finer details of these color requirements and advocating for the flexibility designers need so they can balance when their requirements conflict.

> though we're working on a blue exception, to permit #00a to calculate as #000 for short periods (i.e. link use) as blue is not a significant part of readability luminance contrast

Could you speak more about this? Links and other short interactive elements aren't always blue, so is the exception you are considering only related to blue elements, or any hue?

**Reply by front-endian** [reply]
Wanted to circle back and see if you had any feedback based on the questions above @Myndex. Especially this:
> A LUT to determine what is the acceptable contrast for a given line/element thickness would be extremely helpful for creating accessible data visualizations, non-text interactive elements etc.

I get a lot of questions from designers about designs with lines that are 1px thick (occasionally 2px). The general guidelines listed at https://www.myndex.com/APCA/ are okay, but only give the case of 3px lines. Have a LUT for non-text elements like section dividing lines, icons, input borders, etc would be extremely helpful.

**Reply by Myndex** [reply]
Hi @justcolin 

I've been working on a new iteration of the basic tool, and I've set it up specifically such that I can add in a dynamic non text LUT.

The redesigned tool is:   https://www.myndex.com/APCA/indexK   (temp development link, will expire in a few weeks probably)

While it does not show such a defined LUT yet, you can see that that is on the way, with some use case features now implemented.

Also, did I ever give you a link to this thread?:
https://github.com/Myndex/SAPC-APCA/discussions/39

It discusses use cases and the direction headed for non text, and has some informative graphics.

## Lines in the sand
So, with a look at the contrast sensitive graphic on the above link, we can very "roughly" say that 1px needs a near maximum contrast, as in 90 or more.

**_This is very rough, please don't set it in stone anywhere:_**
1px needs 90, 1.44px needs 75, 2 px needs 60, 3 px needs 45, 4.5px needs 30. 7px needs 15


But context is important... are we talking about a singular 1px or 2px line that has beaning? Or a series of 1px lines that are part of a texture?

**Comment by UXPublishing** [comment]
It has been a few years since I read through all the APCA guidelines. I am trying to update some of the guidelines in my personal spreadsheet tool. I am essentially defining minimum Lc values rules for each font-size and font-weight combination. Like a master look up table that can accommodate pretty much every useful combination. 

A few years ago, I think I just made assumptions to fill in the gaps in the font-sizes in the lookup table. 
For instance if the lookup table defined the following:
font-size 18px and font-weight 400 requires min. Lc 75
font-size 21px and font-weight 400 requires min. Lc 70
I assumed that font-sizes 18px to 20.9px and font-weight 400 required min. Lc 75.

I am wondering if I should make those same kind of assumptions with the Beta lookup tables. I also noticed that Lc values for the body text/block text are only available for certain font-size and font-weight combinations. 

For Non-Body text: 
- Is there a list or complete lookup table that contains all the minimum Lc values for each font-size (every whole pixel increment from 12px to 120px) and font-weight combination? 

For Body text: 
- Is there a list or complete lookup table that contains all the minimum Lc values for each font-size (every whole pixel increment from 12px to 120px) and font-weight combination? 

For example:
font-size: 12px; font-weight: 400; Lc: ?
font-size: 13px; font-weight: 400; Lc: ?
font-size: 14px; font-weight: 400; Lc: 100
font-size: 15px; font-weight: 400; Lc: 100
font-size: 16px; font-weight: 400; Lc: 90
font-size: 17px; font-weight: 400; Lc: ?
font-size: 18px; font-weight: 400; Lc: 75
...
font-size: 95px; font-weight: 400; Lc: ?
font-size: 96px; font-weight: 400; Lc: ?
...
font-size: 119 px; font-weight: 400; Lc: ?
font-size: 120 px; font-weight: 400; Lc: ?


Edit: 
Also, is this lookup table guideline still relevant to the Beta lookup table?
"Many serif fonts should use values for the row above (e.g., Times, Georgia, Cambria, Courier), and especially should be compared in terms of x-height."

I had read through https://github.com/Myndex/SAPC-APCA/discussions/39 but I couldn't find an easy to consume table or list.
I just read through this discussion and I seem to be making similar guidelines sorted by the type of text or non-text elements: https://github.com/Myndex/SAPC-APCA/discussions/62.

**Reply by Myndex** [reply]
Hi @UXPublishing 

The current LUTs are: https://github.com/Myndex/apca-w3/blob/master/data/LUT-GseriesMay28-2022.js

That repo and the NPM also have utilities for transversing the LUTs. Linear interpolation is permitted, as is rounding the LC value to the nearest whole number.

Keep in mind the LUT was created based on the reference font Helvetica. Other fonts will differ, and there's some guidance on that at the [APCA Readability Criterion](https://readtech.org/ARC/tests/visual-readability-contrast/?tn=methods#test-methods).
